src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs

```diff
@@ -56,17 +52,24 @@ protected ServiceHubServiceBase(IServiceProvider serviceProvider, Stream stream)
             Logger = (TraceSource)serviceProvider.GetService(typeof(TraceSource));
             Logger.TraceInformation($"{DebugInstanceString} Service instance created");
 
-            _shutdownCancellationSource = new CancellationTokenSource();
-            ShutdownCancellationToken = _shutdownCancellationSource.Token;
-
             // due to this issue - https://github.com/dotnet/roslyn/issues/16900#issuecomment-277378950
             // all sub type must explicitly start JsonRpc once everything is
             // setup
-            Rpc = new JsonRpc(new JsonRpcMessageHandler(stream, stream), this);
-            Rpc.JsonSerializer.Converters.Add(AggregateJsonConverter.Instance);
+            var jsonFormatter = new JsonMessageFormatter();
+            jsonFormatter.JsonSerializer.Converters.Add(AggregateJsonConverter.Instance);
+
+            Rpc = new JsonRpc(new HeaderDelimitedMessageHandler(stream, jsonFormatter), this);
+            Rpc.CancelLocallyInvokedMethodsWhenConnectionIsClosed = true;
+            Rpc.TraceSource = Logger;
```

Comment:
this pattern seems to be repeated numerous times.  consider extracting into common location?

Reply:
no common location since 2 doesnt have any layer that has common that has reference to JsonRpc

Reply:
Linked file?

Reply:
sure.